aws/protocol/json/decoder_util.go

```diff
@@ -0,0 +1,123 @@
+package json
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+)
+
+// GetErrorInfo util looks for code, __type, and message members in the
+// json body. These members are optionally available, and the function
+// returns the value of member if it is available. This function is useful to
+// identify the error code, msg in a REST JSON error response.
+func GetErrorInfo(decoder *json.Decoder) (string, string, error) {
+
+	var code, typeCode, msg string
+
+	startToken, err := decoder.Token()
+	if err == io.EOF {
+		return "", "", nil
+	}
+	if err != nil {
+		return "", "", err
+	}
+
+	if t, ok := startToken.(json.Delim); !ok || t.String() != "{" {
+		return "", "", fmt.Errorf("expected start token to be {")
+	}
+
+	for decoder.More() {
+		t, err := decoder.Token()
+		if err != nil {
+			return "", "", err
+		}
+
+		switch t {
+		case "code":
+			v, err := decoder.Token()
+			if err != nil {
+				return "", "", err
+			}
+			code = v.(string)
+			break
+		case "message":
+			v, err := decoder.Token()
+			if err != nil {
+				return "", "", err
+			}
+			msg = v.(string)
+			break
+		case "__type":
```

Comment:
decent amount of duplicated code here between these values. can this be collapsed a bit to share some of the logic?  Using a `continue` in the default case will help allow the bottom of the `for` statement to be used for generic logic.  Especially since the `v` decoded is always used as a string.


```go
for decoder.More() {
     var target *string
     switch t {
     case strings.EqualFold("code"):
              target = &code

     case strings.EqualFold("__type"):
              target = &typeCode

     case strings.EqualFold("message"):
              target = &msg

    default:
           DiscardUnknownField(decoder)
           continue
     }

     v, err := decoder.Token()
     if err != nil { return "", "", err }
     
     *target = v.(string)
}